### PR TITLE
OCPBUGS-30860: Use infrastructure API to detect cluster topology

### PR DESCRIPTION
**Pull request by tkashem**

### CODE REVIEW

File: data/data/bootstrap/files/usr/local/bin/wait-for-ha-api.sh [hunk 1]
```diff
@@ -13,6 +27,32 @@ wait_for_ha_api() {
     done
 }
 
+
+get_topology() {
+    output=$(oc --kubeconfig="$KUBECONFIG" get infrastructures cluster -o jsonpath='{.status.infrastructureTopology}' 2>&1 )
```
Comment (patrickdillon) [inline data/data/bootstrap/files/usr/local/bin/wait-for-ha-api.sh:32]: I think we want `controlPlaneTopology` not `infrastructureTopology`

Comment (eggfoobar) [reply data/data/bootstrap/files/usr/local/bin/wait-for-ha-api.sh:32]: Yeah that would be safer, since SNO with workers is a valid configuration

**Comment by eggfoobar** [comment]
/payload-aggregate periodic-ci-openshift-release-master-nightly-4.16-e2e-aws-ovn-single-node-serial 3

**Comment by patrickdillon** [comment]
> /payload-aggregate periodic-ci-openshift-release-master-nightly-4.16-e2e-aws-ovn-single-node-serial 3

Cluster-api based (non-terraform) aws installs landed on Friday. So these tests will use capi. We did not test sno on capi-installer as I did not know it was done outside of bootstrap in place until this came up. But I am pleased to have seen that initial ci tests have already passed. Will be watching this and looking at sno more closely.

**Comment by geliu2016** [comment]
@wangke19 could you please take a view this fix? thanks

File: data/data/bootstrap/files/usr/local/bin/wait-for-ha-api.sh [hunk 2]
```diff
@@ -1,10 +1,24 @@
 #!/usr/bin/env bash
 
 wait_for_ha_api() {
-    if [ "$BOOTSTRAP_INPLACE" = true ]
-    then
-	return 0
-    fi
+    while :
```
Comment (Miciah) [inline data/data/bootstrap/files/usr/local/bin/wait-for-ha-api.sh:4]: Mixing tabs and spaces makes this difficult to read.  Better to use spaces here to be consistent with the rest of the file.

File: data/data/bootstrap/files/usr/local/bin/wait-for-ha-api.sh [hunk 3]
```diff
@@ -13,6 +27,32 @@ wait_for_ha_api() {
...
+	return 1
+    fi
+
+    if [ -z "$output" ]
+    then
+	echo "status.infrastructureTopology of the infrastructures/cluster object is empty"
+	return 1
+    fi
+
+    if [ "$output" == "SingleReplica" ]
```
Comment (Miciah) [inline data/data/bootstrap/files/usr/local/bin/wait-for-ha-api.sh:48]: This mix of Bourne syntax (`[`) and Bash syntax (`==`) makes my eye twitch.  * grin *.  Might as well use `[[` here to be consistent with the use of `[[` elsewhere in the file and the use of Bash-specific `==`.

**Comment by tkashem** [comment]
sno run failed: https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_installer/8429/pull-ci-openshift-installer-master-e2e-aws-ovn-single-node/1792919444352143360 (unrelated, i believe)

/test e2e-aws-ovn-single-node

**Comment by patrickdillon** [comment]
/approve

The approach looks good to me, but do we want to pursue this or https://github.com/openshift/cluster-kube-apiserver-operator/pull/1678 + https://github.com/openshift/installer/pull/8404?

Ultimately I'm more comfortable with the operator approach, but I'm not sure if the time frame works out

**Comment by tkashem** [comment]
Yes, +1 for https://github.com/openshift/cluster-kube-apiserver-operator/pull/1678, and this PR is a short term solution to unblock the SNO team

**Comment by sadasu** [comment]
/override ci/prow/okd-images

Known issue.

File: data/data/bootstrap/files/usr/local/bin/wait-for-ha-api.sh [hunk 4]
```diff
@@ -1,18 +1,58 @@
...
     while ! is_api_available
     do
-	sleep 5
+        sleep 5
     done
 }
 
+
+get_topology() {
+    output=$(oc --kubeconfig="$KUBECONFIG" get infrastructures cluster -o jsonpath='{.status.controlPlaneTopology}' 2>&1 )
```
Comment (tkashem) [inline data/data/bootstrap/files/usr/local/bin/wait-for-ha-api.sh:32]: @sadasu this is the proper way to identiy single node topology, and this PR will unblock the SNO team

**Comment by sadasu** [comment]
LGTM
Waiting for ci/prow/e2e-aws-ovn-single-node  to complete

**Comment by sadasu** [comment]
/override ci/prow/e2e-aws-ovn
probably broken because of kube rebase.

**Pull request merged**